File: app/src/main/java/uk/co/victoriajanedavis/chatapp/data/repositories/cache/MessageCache.java
package uk.co.victoriajanedavis.chatapp.data.repositories.cache;

import java.util.Date;
import java.util.List;
import java.util.UUID;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope;
import uk.co.victoriajanedavis.chatapp.data.model.db.MessageDbModel;
import uk.co.victoriajanedavis.chatapp.data.room.ChatAppDatabase;
import uk.co.victoriajanedavis.chatapp.data.room.daos.MessageDao;
import uk.co.victoriajanedavis.chatapp.domain.Cache.DiskCache;

@ApplicationScope
public class MessageCache implements DiskCache<UUID, MessageDbModel> {

    private final MessageDao dao;


    @Inject
    public MessageCache(ChatAppDatabase database) {
        this.dao = database.messageDao();
    }

    @Override
    public void putSingular(@NonNull MessageDbModel messageDbModel) {
        dao.insertMessage(messageDbModel);
    }

    @Override
    public void putAll(@NonNull List<MessageDbModel> messageDbModels) {
        dao.insertMessages(messageDbModels);
    }

    @Override
    public void replaceAll(@Nullable UUID chatUuid, @NonNull List<MessageDbModel> messageDbModels) {
        if (chatUuid == null) dao.replaceAll(messageDbModels);
        else dao.replaceAllByChat(chatUuid, messageDbModels);
    }

    @Override
    public void delete(@NonNull MessageDbModel messageDbModel) {
        dao.deleteMessage(messageDbModel);
    }

    @Override
    public void clear() {
        dao.clear();
    }

    @NonNull
    @Override
    public Observable<MessageDbModel> getSingular(@NonNull UUID uuid) {
        return dao.get(uuid).toObservable();
    }

    @NonNull
    @Override
    public Observable<List<MessageDbModel>> getAll(@Nullable UUID chatUuid) {
        return dao.getMessagesByChatUuid(chatUuid).toObservable();
    }

    @NonNull
    public Single<Date> getDateOfOldestMessageByChat(@Nullable UUID chatUuid) {
        return dao.getDateOfOldestMessageByChat(chatUuid);
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/data/repositories/cache/MessageCache.kt
package uk.co.victoriajanedavis.chatapp.data.repositories.cache

import java.util.Date
import java.util.UUID

import javax.inject.Inject

import io.reactivex.Observable
import io.reactivex.Single
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope
import uk.co.victoriajanedavis.chatapp.data.model.db.MessageDbModel
import uk.co.victoriajanedavis.chatapp.data.room.ChatAppDatabase
import uk.co.victoriajanedavis.chatapp.data.room.daos.MessageDao
import uk.co.victoriajanedavis.chatapp.domain.Cache.DiskCache

@ApplicationScope
class MessageCache @Inject constructor(
    private val dao: MessageDao
) : DiskCache<UUID, MessageDbModel> {


    override fun putSingular(value: MessageDbModel) {
        dao.insertMessage(value)
    }

    override fun putAll(valuesList: List<MessageDbModel>) {
        dao.insertMessages(valuesList)
    }

    // Key should be chatUuid
    override fun replaceAll(key: UUID?, valuesList: List<MessageDbModel>) {
        if (key == null)
            dao.replaceAll(valuesList)
        else
            dao.replaceAllByChat(key, valuesList)
    }

    override fun delete(value: MessageDbModel) {
        dao.deleteMessage(value)
    }

    override fun clear() {
        dao.clear()
    }

    override fun getSingular(key: UUID): Observable<MessageDbModel> {
        return dao.get(key).toObservable()
    }

    // Key should be chatUuid
    override fun getAll(key: UUID?): Observable<List<MessageDbModel>> {
        return dao.getMessagesByChatUuid(key!!)
            .distinctUntilChanged()
            .toObservable()
    }

    fun getDateOfOldestMessageByChat(chatUuid: UUID): Single<Date> {
        return dao.getDateOfOldestMessageByChat(chatUuid)
    }
}
